pkg/repository/machine.go

```diff
@@ -0,0 +1,480 @@
+package repository
+
+import (
+	"context"
+	"slices"
+	"strings"
+
+	"github.com/hibiken/asynq"
+	"github.com/metal-stack/api/go/enum"
+	apiv2 "github.com/metal-stack/api/go/metalstack/api/v2"
+	"github.com/metal-stack/metal-apiserver/pkg/db/generic"
+	"github.com/metal-stack/metal-apiserver/pkg/db/metal"
+	"github.com/metal-stack/metal-apiserver/pkg/db/queries"
+	"github.com/metal-stack/metal-apiserver/pkg/errorutil"
+	"github.com/metal-stack/metal-lib/pkg/pointer"
+	"github.com/metal-stack/metal-lib/pkg/tag"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	machineRepository struct {
+		s     *Store
+		scope *ProjectScope
+	}
+)
+
+func (r *machineRepository) get(ctx context.Context, id string) (*metal.Machine, error) {
+	machine, err := r.s.ds.Machine().Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	return machine, nil
+}
+
+func (r *machineRepository) matchScope(machine *metal.Machine) bool {
+	if r.scope == nil {
+		return true
+	}
+
+	if machine.Allocation == nil {
+		return true
+	}
+
+	return r.scope.projectID == pointer.SafeDeref(machine).Allocation.Project
+}
+
+func (r *machineRepository) create(ctx context.Context, req *apiv2.MachineServiceCreateRequest) (*metal.Machine, error) {
+	panic("unimplemented")
+}
+
+func (r *machineRepository) update(ctx context.Context, m *metal.Machine, req *apiv2.MachineServiceUpdateRequest) (*metal.Machine, error) {
+	if m.Allocation == nil {
+		return m, errorutil.InvalidArgument("only allocated machines can be updated")
+	}
+
+	if req.Description != nil {
+		m.Allocation.Description = *req.Description
+	}
+
+	if req.Labels != nil {
+		m.Allocation.Labels = updateLabelsOnMap(req.Labels, m.Allocation.Labels)
+	}
+
+	if len(req.SshPublicKeys) > 0 && m.Allocation != nil {
+		m.Allocation.SSHPubKeys = req.SshPublicKeys
+	}
+	if err := r.s.ds.Machine().Update(ctx, m); err != nil {
+		return nil, errorutil.Convert(err)
+	}
+
+	return m, nil
+}
+
+func (r *machineRepository) delete(ctx context.Context, m *metal.Machine) error {
+	var (
+		uuid           *string
+		allocationUUID *string
+	)
+	uuid = &m.ID
+	if m.Allocation != nil {
+		uuid = nil
+		allocationUUID = &m.Allocation.UUID
+	}
+	info, err := r.s.async.NewMachineDeleteTask(uuid, allocationUUID)
+	if err != nil {
+		return err
+	}
+
+	r.s.log.Info("machine delete queued", "info", info)
+
+	return nil
+}
+
+func (r *machineRepository) find(ctx context.Context, rq *apiv2.MachineQuery) (*metal.Machine, error) {
+	ip, err := r.s.ds.Machine().Find(ctx, r.scopedMachineFilters(queries.MachineFilter(rq))...)
+	if err != nil {
+		return nil, err
+	}
+
+	return ip, nil
+}
+
+func (r *machineRepository) list(ctx context.Context, rq *apiv2.MachineQuery) ([]*metal.Machine, error) {
+	machines, err := r.s.ds.Machine().List(ctx, r.scopedMachineFilters(queries.MachineFilter(rq))...)
+	if err != nil {
+		return nil, err
+	}
+
+	slices.SortFunc(machines, func(a, b *metal.Machine) int {
+		return strings.Compare(a.ID, b.ID)
+	})
```

Comment:
Possible improvement for follow-up: Add `sort` to the interface and run this for all repo entities in the generic find.